core/src/test/scala/unit/kafka/log/LogTest.scala

```diff
@@ -1013,6 +1013,61 @@ class LogTest extends JUnitSuite {
   }
 
   @Test
+  def testOverCompactedLogRecovery(): Unit = {
+    // append some messages to create some segments
+    val logProps = new Properties()
+    logProps.put(LogConfig.SegmentBytesProp, 1000: java.lang.Integer)
+    logProps.put(LogConfig.MaxMessageBytesProp, 64*1024: java.lang.Integer)
+    logProps.put(LogConfig.IndexIntervalBytesProp, 1: java.lang.Integer)
+    val config = LogConfig(logProps)
+    val log = new Log(logDir,
+      config,
+      recoveryPoint = 0L,
+      time.scheduler,
+      time)
+    val set1 = new ByteBufferMessageSet(
+      compressionCodec = NoCompressionCodec,
+      offsetSeq = Seq(0),
+      messages = new Message("v1".getBytes(), "k1".getBytes(), Message.NoTimestamp, magicValue = Message.CurrentMagicValue))
+    val set2 = new ByteBufferMessageSet(
+      compressionCodec = NoCompressionCodec,
+      offsetSeq = Seq(Integer.MAX_VALUE.toLong + 1),
+      messages = new Message("v2".getBytes(), "k2".getBytes(), Message.NoTimestamp, magicValue = Message.CurrentMagicValue))
+    val set3 = new ByteBufferMessageSet(
+      compressionCodec = NoCompressionCodec,
+      offsetSeq = Seq(Integer.MAX_VALUE.toLong + 2),
+      messages = new Message("v3".getBytes(), "k3".getBytes(), Message.NoTimestamp, magicValue = Message.CurrentMagicValue))
+    val set4 = new ByteBufferMessageSet(
+      compressionCodec = NoCompressionCodec,
+      offsetSeq = Seq(Integer.MAX_VALUE.toLong + 3),
+      messages = new Message("v4".getBytes(), "k4".getBytes(), Message.NoTimestamp, magicValue = Message.CurrentMagicValue))
+
+    //Writes into an empty log with baseOffset 0
+    log.append(set1, false)
+
+    //This write will roll the segment, and create a new segment with baseOffset = log end of previous segment  1
+    log.append(set2, false)
+
+    //This write will also roll the segment, ""                                       "" Integer.MAX_VALUE + 2
+    log.append(set3, false)
+
+    //Since the last segment got baseOffset = Integer.MAX_VALUE + 2, this set does not roll the log.
+    log.append(set4, false)
+
+    log.close()
+    val indexFiles = logDir.listFiles(new FilenameFilter {
```

Comment:
Nit: it's fine to simply do `logDir.listFiles.asScala.filter(..)` instead of passing a `FilenameFilter`.